applications/FluidDynamicsApplication/custom_elements/two_fluid_navier_stokes.cpp

```diff
@@ -1879,15 +1926,240 @@ void TwoFluidNavierStokes<TElementData>::ComputeSplitting(
+        if (rData.Distance[i] > 0.0){
+            enr_neg_interp(i, i) = 1.0;
+        } else{
+            enr_pos_interp(i, i) = 1.0;
+        }
+    }
+
+    // Call the Interface negative side shape functions calculator
+    pModifiedShapeFunctions->ComputeInterfaceNegativeSideShapeFunctionsAndGradientsValues(
+        rInterfaceShapeFunctionNeg,
+        rInterfaceShapeDerivativesNeg,
+        rInterfaceWeightsNeg,
+        GeometryData::GI_GAUSS_2);
+
+    // Call the Interface negative side normal functions calculator
+    pModifiedShapeFunctions->ComputeNegativeSideInterfaceAreaNormals(
+        rInterfaceNormalsNeg,
+        GeometryData::GI_GAUSS_2);
+
+    for (unsigned int gp = 0; gp < rInterfaceNormalsNeg.size(); ++gp){
+        const double normal_norm = norm_2(rInterfaceNormalsNeg[gp]);
+        rInterfaceNormalsNeg[gp] /= normal_norm;
+    }
+
+    // Compute the enrichment shape function values at the interface gauss points using the enrichment interpolation matrices
+    rEnrInterfaceShapeFunctionPos = prod(rInterfaceShapeFunctionNeg, enr_pos_interp);
+    rEnrInterfaceShapeFunctionNeg = prod(rInterfaceShapeFunctionNeg, enr_neg_interp);
+}
+
+template <>
+ModifiedShapeFunctions::UniquePointer TwoFluidNavierStokes< TwoFluidNavierStokesData<2, 3> >::ModifiedShapeFunctionsUtility(
+    const GeometryType::Pointer pGeometry,
+    const Vector& rDistances)
+{
+    auto p_modified_sh_func = Kratos::make_unique<Triangle2D3ModifiedShapeFunctions>(pGeometry, rDistances);
+    return p_modified_sh_func;
+}
+
+template <>
+ModifiedShapeFunctions::UniquePointer TwoFluidNavierStokes< TwoFluidNavierStokesData<3, 4> >::ModifiedShapeFunctionsUtility(
+        const GeometryType::Pointer pGeometry,
+        const Vector& rDistances)
+{
+    auto p_modified_sh_func = Kratos::make_unique<Tetrahedra3D4ModifiedShapeFunctions>(pGeometry, rDistances);
+    return p_modified_sh_func;
+}
+
+template <class TElementData>
+void TwoFluidNavierStokes<TElementData>::CalculateCurvature(
+        const Matrix& rInterfaceShapeFunctions,
+        Vector& rInterfaceCurvature)
+{
+    auto geom = this->GetGeometry();
+    const unsigned int n_gpt = rInterfaceShapeFunctions.size1();
+
+    rInterfaceCurvature.resize(n_gpt, false);
+
+    for (unsigned int gpt = 0; gpt < n_gpt; ++gpt){
+        double curvature = 0.0;
+        for (unsigned int i = 0; i < NumNodes; ++i){
+            curvature += rInterfaceShapeFunctions(gpt,i) * geom[i].GetValue(CURVATURE);
+        }
+        rInterfaceCurvature[gpt] = curvature;
+    }
+}
+
+template <class TElementData>
+void TwoFluidNavierStokes<TElementData>::SurfaceTension(
+    const double coefficient,
+    const Vector& rCurvature,
+    const Vector& rInterfaceWeights,
+    const Matrix& rInterfaceShapeFunctions,
+    const std::vector<Vector>& rInterfaceNormalsNeg,
+    VectorType& rRHS)
+{
+    VectorType rhs = ZeroVector(NumNodes*NumNodes); //Size is NumNodes x (NumDim+1)
```

Comment:
Why allocating this auxiliary vector? I think you can directly add to the `rRHS`.

Reply:
Changed to:
rRHS[ j*(NumNodes) + dim ] -= ...